Fix concurrent map iteration and map write panic

thedadams opened the pull request:
There is a strange panic on Rancher startup where a concurrent map iteration and
 map write panic can happen in the nodesyncer.

 This change will pass deep copies to the functions where this has been seen to
 ensure that the iterations are happening on a different map to get around the
 concurrent panic.

 The DeepCopy was already happening in these cases, but it was happening after
 the iteration. Now it happens before.

Oats87 reviewed (approved):
lgtm

Oats87 commented:
Using the cherry-picked commit, I still saw an instance of the concurrent map r/w panic.

```
fatal error: concurrent map iteration and map write

goroutine 4633 [running]:
runtime.throw(0x45ee78d, 0x26)
	/usr/lib64/go/1.16/src/runtime/panic.go:1117 +0x72 fp=0xc006405678 sp=0xc006405648 pc=0x4386b2
runtime.mapiternext(0xc006405780)
	/usr/lib64/go/1.16/src/runtime/map.go:858 +0x54c fp=0xc0064056f8 sp=0xc006405678 pc=0x410e4c
k8s.io/apimachinery/pkg/apis/meta/v1.(*ObjectMeta).DeepCopyInto(0xc0004ac320, 0xc008370c20)
	/go/src/github.com/rancher/rancher/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/zz_generated.deepcopy.go:683 +0x5f3 fp=0xc0064057f0 sp=0xc0064056f8 pc=0xa43e93
k8s.io/api/core/v1.(*Node).DeepCopyInto(0xc0004ac300, 0xc008370c00)
	/go/src/github.com/rancher/rancher/vendor/k8s.io/api/core/v1/zz_generated.deepcopy.go:2336 +0x9a fp=0xc006405818 sp=0xc0064057f0 pc=0xc4239a
k8s.io/api/core/v1.(*Node).DeepCopy(...)
	/go/src/github.com/rancher/rancher/vendor/k8s.io/api/core/v1/zz_generated.deepcopy.go:2348
github.com/rancher/rancher/pkg/controllers/managementuser/nodesyncer.(*nodesSyncer).syncLabels(0xc00326ee60, 0xc000c181c8, 0x13, 0xc0002edb00, 0x10, 0x7fde05347c28, 0x10, 0xc0023bbb10)
	/go/src/github.com/rancher/rancher/pkg/controllers/managementuser/nodesyncer/nodessyncer.go:255 +0x18d fp=0xc0064059f0 sp=0xc006405818 pc=0x2b3eecd
github.com/rancher/rancher/pkg/controllers/managementuser/nodesyncer.(*nodesSyncer).syncLabels-fm(0xc000c181c8, 0x13, 0xc0002edb00, 0x10, 0x3ae3820, 0x1, 0xc0023bbb10)
	/go/src/github.com/rancher/rancher/pkg/controllers/managementuser/nodesyncer/nodessyncer.go:240 +0x51 fp=0xc006405a40 sp=0xc0064059f0 pc=0x2b46611
github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3.(*nodeController).AddHandler.func1(0xc000c181c8, 0x13, 0x44b77e0, 0xc0002edb00, 0xc006405b58, 0x3, 0x3, 0x0)
	/go/src/github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3/zz_generated_node_controller.go:156 +0x67 fp=0xc006405a88 sp=0xc006405a40 pc=0x16e9067
github.com/rancher/norman/controller.(*genericController).AddHandler.func1(0xc000c181c8, 0x13, 0x4d06f30, 0xc0002edb00, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/rancher/rancher/vendor/github.com/rancher/norman/controller/generic_controller.go:62 +0x224 fp=0xc006405b98 sp=0xc006405a88 pc=0x142ca04
github.com/rancher/lasso/pkg/controller.SharedControllerHandlerFunc.OnChange(0xc00439f140, 0xc000c181c8, 0x13, 0x4d06f30, 0xc0002edb00, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/rancher/rancher/vendor/github.com/rancher/lasso/pkg/controller/sharedcontroller.go:29 +0x4e fp=0xc006405be8 sp=0xc006405b98 pc=0x13ecdce
github.com/rancher/lasso/pkg/controller.(*SharedHandler).OnChange(0xc002422f00, 0xc000c181c8, 0x13, 0x4d06f30, 0xc0002edb00, 0xc000c94f01, 0x0)
	/go/src/github.com/rancher/rancher/vendor/github.com/rancher/lasso/pkg/controller/sharedhandler.go:69 +0x14c fp=0xc006405d08 sp=0xc006405be8 pc=0x13ef02c
github.com/rancher/lasso/pkg/controller.(*controller).syncHandler(0xc0008a1600, 0xc000c181c8, 0x13, 0xc000c95048, 0x0)
	/go/src/github.com/rancher/rancher/vendor/github.com/rancher/lasso/pkg/controller/controller.go:218 +0xd1 fp=0xc006405d58 sp=0xc006405d08 pc=0x13ec191
github.com/rancher/lasso/pkg/controller.(*controller).processSingleItem(0xc0008a1600, 0x3ae49a0, 0xc007cc7f30, 0x0, 0x0)
	/go/src/github.com/rancher/rancher/vendor/github.com/rancher/lasso/pkg/controller/controller.go:200 +0xe7 fp=0xc006405e10 sp=0xc006405d58 pc=0x13ebd67
github.com/rancher/lasso/pkg/controller.(*controller).processNextWorkItem(0xc0008a1600, 0x203001)
	/go/src/github.com/rancher/rancher/vendor/github.com/rancher/lasso/pkg/controller/controller.go:177 +0x54 fp=0xc006405e58 sp=0xc006405e10 pc=0x13ebb54
github.com/rancher/lasso/pkg/controller.(*controller).runWorker(...)
	/go/src/github.com/rancher/rancher/vendor/github.com/rancher/lasso/pkg/controller/controller.go:166
github.com/rancher/lasso/pkg/controller.(*controller).runWorker-fm()
	/go/src/github.com/rancher/rancher/vendor/github.com/rancher/lasso/pkg/controller/controller.go:165 +0x3f fp=0xc006405e80 sp=0xc006405e58 pc=0x13f0d5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc0046688a0)
	/go/src/github.com/rancher/rancher/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x5f fp=0xc006405ed0 sp=0xc006405e80 pc=0xc904df
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0046688a0, 0x4cddd60, 0xc000f6b3e0, 0x1, 0xc00181e000)
	/go/src/github.com/rancher/rancher/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0x9b fp=0xc006405f58 sp=0xc006405ed0 pc=0xc8ed5b
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0046688a0, 0x3b9aca00, 0x0, 0x47b6f01, 0xc00181e000)
	/go/src/github.com/rancher/rancher/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x98 fp=0xc006405f90 sp=0xc006405f58 pc=0xc8ec98
k8s.io/apimachinery/pkg/util/wait.Until(0xc0046688a0, 0x3b9aca00, 0xc00181e000)
	/go/src/github.com/rancher/rancher/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x4d fp=0xc006405fc8 sp=0xc006405f90 pc=0xc8ebed
runtime.goexit()
	/usr/lib64/go/1.16/src/runtime/asm_amd64.s:1371 +0x1 fp=0xc006405fd0 sp=0xc006405fc8 pc=0x472641
created by github.com/rancher/lasso/pkg/controller.(*controller).run
	/go/src/github.com/rancher/rancher/vendor/github.com/rancher/lasso/pkg/controller/controller.go:134 +0x35b

```

lazybetrayer commented:
@Oats87 can you try my PR https://github.com/rancher/rancher/pull/36130?

thedadams commented:
@lazybetrayer Thanks for the posting your potential fix. I updated mine to be an optimization of yours. If you update your PR (and this fix works), then we will be sure to merge your PR instead of mine. I updated mine for ease.

@Oats87 Try this commit and see if it continues to happen.

Oats87 reviewed (approved):
Have not seen a startup panic in ~10 Rancher startups, whereas normally I get 1/3 startups to panic.

lazybetrayer commented on pkg/controllers/managementuser/nodesyncer/nodessyncer.go:
```diff
@@ -706,8 +706,15 @@ func cleanStatus(machine *v3.Node) {
 
-	delete(machine.Status.NodeAnnotations, podresources.RequestsAnnotation)
-	delete(machine.Status.NodeAnnotations, podresources.LimitsAnnotation)
+	annoMap := make(map[string]string, len(machine.Status.NodeAnnotations))
+	for key, val := range machine.Status.NodeAnnotations {
+		if key == podresources.LimitsAnnotation || key == podresources.RequestsAnnotation {
+			continue
+		}
+		annoMap[key] = val
+	}
+
+	machine.Status.NodeAnnotations = annoMap
```
I think this is not needed, `machine.Status.NodeAnnotations` is already made a deepcopy in `convertNodeToMachine`

thedadams replied:
The issue is that someone could use the `cleanStatus` function outside of the `convertNodeToMachine` and we would be back in the same position. Just trying to protect against this from happening in the future.

lazybetrayer replied:
What about `machine.Status.Conditions`? It's also modified in this function.

thedadams replied:
Conditions is an slice. I don't think slices have the same problem with iteration/write panics.
